### ISSUE TITLE
test_defaultGAI (twisted.internet.test.test_endpoints.HostnameEndpointsOneIPv4Tests) fails on Python 3.4+

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#7886]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)| @hawkowl reported|
|-|-|
|Trac ID|trac#7886|
|Type|defect|
|Created|2015-05-04 11:24:33Z|
|Branch|https://github.com/twisted/twisted/tree/defaultgai-py34-7886|

See https://buildbot.twistedmatrix.com/builders/debian7-py3.4/builds/4/steps/shell/logs/stdio

```

======================================================================
FAIL: test_defaultGAI (twisted.internet.test.test_endpoints.HostnameEndpointsOneIPv4Tests)
test_defaultGAI
----------------------------------------------------------------------
Traceback (most recent call last):
  File "/buildslave/python-3.4-tests/Twisted/twisted/internet/defer.py", line 150, in maybeDeferred
    result = f(*args, **kw)
  File "/buildslave/python-3.4-tests/Twisted/twisted/internet/utils.py", line 201, in runWithWarningsSuppressed
    reraise(exc_info[1], exc_info[2])
  File "/buildslave/python-3.4-tests/Twisted/twisted/python/compat.py", line 368, in reraise
    raise exception.with_traceback(traceback)
  File "/buildslave/python-3.4-tests/Twisted/twisted/internet/utils.py", line 197, in runWithWarningsSuppressed
    result = f(*a, **kw)
  File "/buildslave/python-3.4-tests/Twisted/twisted/internet/test/test_endpoints.py", line 1659, in test_defaultGAI
    self.assertEqual(ep._getaddrinfo, socket.getaddrinfo)
  File "/buildslave/python-3.4-tests/Twisted/twisted/trial/_synctest.py", line 437, in assertEqual
    super(_Assertions, self).assertEqual(first, second, msg)
  File "/pyenv/versions/3.4.3/lib/python3.4/unittest/case.py", line 797, in assertEqual
    assertion_func(first, second, msg=msg)
  File "/pyenv/versions/3.4.3/lib/python3.4/unittest/case.py", line 790, in _baseAssertEqual
    raise self.failureException(msg)
twisted.trial.unittest.FailTest: <bound method HostnameEndpoint.getaddrinfo[70 chars]b70>> != <function getaddrinfo at 0x7f8b02e9ce18>
```

This is a blocker along with [#7885](https://github.com/twisted/twisted/issues/7885) for putting 3.4 and 3.5 on the supported list. It appears to be fixable by making it a staticmethod.

<details><summary>Searchable metadata</summary>

```
trac-id__7886 7886
type__defect defect
reporter__hawkowl hawkowl
priority__normal normal
milestone__Python_3_x Python-3.x
branch__branches_defaultgai_py34_7886 branches/defaultgai-py34-7886
branch_author__hawkowl hawkowl
status__closed closed
resolution__fixed fixed
component__core core
keywords__ 
time__1430738673055258 1430738673055258
changetime__1430977505939865 1430977505939865
version__None None
owner__hawkowl hawkowl

```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_1"></a>|@hawkowl commented|
|-|-|

(In [44662]) Branching to defaultgai-py34-7886.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_2"></a>|@hawkowl commented|
|-|-|

This manifests in HostnameEndpoint being broken on 3.4+, here's a script to test:

```
from twisted.internet import endpoints, task, protocol, defer
from twisted.protocols.basic import LineReceiver

class LC(LineReceiver):

    def connectionMade(self):
        self.transport.write(b"GET / HTTP/1.1\r\nHost: www.google.com\r\nConnection: close\r\n\r\n")

    def lineReceived(self, line):
        print(line)

    def connectionLost(self, reason):
        self.reactor.stop()


@defer.inlineCallbacks
def main(reactor):

    h = endpoints.HostnameEndpoint(reactor, "google.com", 80)
    d = yield h.connect(protocol.Factory.forProtocol(LC))
    d.reactor = reactor
    yield defer.Deferred()

task.react(main)
```

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_3"></a>|@hawkowl commented|
|-|-|

Fix done, test now passes, and my example script passes. I am not sure how to integrate the example script as a test, because we don't actually want to call gai.

The test now passes on https://buildbot.twistedmatrix.com/builders/debian7-py3.4/builds/5/steps/shell/logs/stdio and the rest of the builders are spun.

Please review.

**Comment by twisted-trac** [comment]
|[<img alt="glyph's avatar" src="https://avatars.githubusercontent.com/u/716529?s=50" width="50" height="50">](https://github.com/glyph)<a name="note_4"></a>|@glyph set owner to @hawkowl|
|-|-|

Simple enough; please land.

**Comment by twisted-trac** [comment]
|[<img alt="hawkowl's avatar" src="https://avatars.githubusercontent.com/u/3307100?s=50" width="50" height="50">](https://github.com/hawkowl)<a name="note_5"></a>|@hawkowl set status to `closed`|
|-|-|

(In [44669]) Merge defaultgai-py34-7886: twisted.internet.endpoints.HostnameEndpoint now uses getaddrinfo properly on Python 3.4 and above.

Author: hawkowl
Reviewer: glyph
Fixes: [#7886](https://github.com/twisted/twisted/issues/7886)
